src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -836,8 +836,9 @@ public void recover(RecoveryHandler recoveryHandler) throws EngineException {
             maybeFailEngine("recovery phase 3", e);
             throw new RecoveryEngineException(shardId, 3, "Execution failed", wrapIfClosed(e));
         } finally {
-            Releasables.close(success, onGoingRecoveries, writeLock, phase1Snapshot,
-                    phase2Snapshot, phase3Snapshot); // hmm why can't we use try-with here?
+            // close the snapshots first to release the reference to the translog file, so a flush post recovery can delete it
+            Releasables.close(success, phase1Snapshot, phase2Snapshot, phase3Snapshot,
+                    onGoingRecoveries, writeLock); // hmm why can't we use try-with here?
```

Comment:
we can do try-with but we need to have 2 try blocks. since the write lock needs to be released last. but in a try / with  blokc it's released before the finally block is executed